File: Wayfork/WayforkCore/Sources/WayforkCore/SingBox/RuleSetGenerator.swift
import Foundation

/// Produces `rules-t-<id>.json` and `rules-direct.json` contents, plus the route-only
/// `…-ip.json` twins holding the IP rules (docs/design/03-routing.md, "Rule-set files").
public enum RuleSetGenerator {
    public static let version = 3

    /// Tag and file of the Direct rule-set: user exceptions plus the built-in local names.
    public static let directTag = "rules-direct"
    public static let directFileName = "rules-direct.json"
    /// Tag and file of the Direct IP rule-set (F11).
    public static let directIPTag = "rules-direct-ip"
    public static let directIPFileName = "rules-direct-ip.json"

    /// Built-in exceptions: names that must never leave the local network (F8).
    public static let builtInDirectSuffixes = [
        ".local", ".lan", ".internal", ".home.arpa", ".localhost",
    ]
    public static let builtInDirectDomains = ["localhost"]

    /// Two rule-set files per tunnel (domains + apps, IPs) plus the two Direct files: file
    /// name → JSON text. Every tunnel in `tunnels` gets its files, even with no rules, and
    /// the Direct files are always present, so the main config stays unchanged when rules
    /// come and go.
    public static func generate(
        tunnels: [Tunnel], activeRules: [UUID: [Rule]], exceptions: [Rule] = []
    ) -> [String: String] {
        var files: [String: String] = [
            directFileName: renderDirect(exceptions: exceptions),
            directIPFileName: renderIP(rules: exceptions),
        ]
        for tunnel in tunnels {
            let rules = activeRules[tunnel.id] ?? []
            files[tunnel.ruleSetFileName] = render(rules: rules)
            files[tunnel.ipRuleSetFileName] = renderIP(rules: rules)
        }
        return files
    }

    /// `…-ip.json` (F11): the group's IP rules as one `ip_cidr` rule, each range minus the
    /// reserved ranges a wide pattern may overlap. Never referenced by DNS rules — there
    /// sing-box would match `ip_cidr` against the answer and skip the domain rules.
    public static func renderIP(rules: [Rule]) -> String {
        let ranges = rules.filter(\.isIP)
            .compactMap { IPv4Prefix($0.pattern) }
            .flatMap { $0.subtracting(all: RulePattern.reservedRanges) }
            .map(\.description)
        let document: [String: Any] = [
            "version": version,
            "rules": ranges.isEmpty ? [] : [["ip_cidr": ranges]],
        ]
        return JSONText.render(document)
    }

    /// Rule-set JSON for one tunnel's rules: one rule object for the domain items and, when
    /// there are app rules (F10), a second one with `process_path_regex` — sing-box ANDs
    /// items of different kinds inside one rule and ORs the rules of a set.
    public static func render(rules: [Rule]) -> String {
        render(rules: rules, domains: [], suffixes: [])
    }

    /// `rules-direct.json`: the built-in local names first, then the user's exceptions.
    public static func renderDirect(exceptions: [Rule]) -> String {
        render(rules: exceptions, domains: builtInDirectDomains, suffixes: builtInDirectSuffixes)
    }

    private static func render(rules: [Rule], domains: [String], suffixes: [String]) -> String {
        var domain = domains
        var domainSuffix = suffixes
        var domainRegex: [String] = []
        var processPathRegex: [String] = []
        for rule in rules {
            switch rule.match {
            case .exact:
                domain.append(rule.pattern)
            case .suffix:
                domain.append(rule.pattern)
                domainSuffix.append("." + rule.pattern)
            case .wildcard:
                domainRegex.append(RulePattern.wildcardRegex(rule.pattern))
            case .app:
                processPathRegex.append(RulePattern.appPathRegex(rule.pattern))
            case .ip:
                continue  // `renderIP`
            }
        }
        var domainRule: [String: Any] = [:]
        if !domain.isEmpty { domainRule["domain"] = domain }
        if !domainSuffix.isEmpty { domainRule["domain_suffix"] = domainSuffix }
        if !domainRegex.isEmpty { domainRule["domain_regex"] = domainRegex }
        var ruleObjects: [[String: Any]] = []
        if !domainRule.isEmpty { ruleObjects.append(domainRule) }
        if !processPathRegex.isEmpty {
            ruleObjects.append(["process_path_regex": processPathRegex])
        }
        let document: [String: Any] = [
            "version": version,
            "rules": ruleObjects,
        ]
        return JSONText.render(document)
    }
}

/// Deterministic JSON rendering for generated files: sorted keys, two-space indent.
enum JSONText {
    static func render(_ object: [String: Any]) -> String {
        // JSONSerialization cannot fail for the plain String/Int/Bool/Array/Dictionary
        // trees the generators build; a failure here would be a programming error.
        let data =
            (try? JSONSerialization.data(
                withJSONObject: object,
                options: [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes])) ?? Data()
        return String(decoding: data, as: UTF8.self) + "\n"
    }
}

File: WayforkWindows/app/lib/core/singbox/rule_set_generator.dart
import 'package:wayfork/core/json_text.dart';
import 'package:wayfork/core/model/rule.dart';
import 'package:wayfork/core/model/tunnel.dart';
import 'package:wayfork/core/platform.dart';
import 'package:wayfork/core/rules/rule_pattern.dart';
import 'package:wayfork/core/support/ipv4_prefix.dart';

/// Produces domain/application and route-only IP rule-set files.
abstract final class RuleSetGenerator {
  static const version = 3;

  /// Tag and file of the Direct rule-set: user exceptions plus local names.
  static const directTag = 'rules-direct';
  static const directFileName = 'rules-direct.json';

  /// Tag and file of the Direct IP rule-set.
  static const directIPTag = 'rules-direct-ip';
  static const directIPFileName = 'rules-direct-ip.json';

  /// Built-in exceptions: names that must never leave the local network.
  static const builtInDirectSuffixes = [
    '.local',
    '.lan',
    '.internal',
    '.home.arpa',
    '.localhost',
  ];
  static const builtInDirectDomains = ['localhost'];

  /// Emits two files per tunnel plus the two Direct files, even when empty.
  static Map<String, String> generate({
    required List<Tunnel> tunnels,
    required Map<String, List<Rule>> activeRules,
    List<Rule> exceptions = const [],
    WayforkPlatform platform = WayforkPlatform.windows,
  }) {
    final files = <String, String>{
      directFileName: renderDirect(exceptions: exceptions, platform: platform),
      directIPFileName: renderIP(exceptions),
    };
    for (final tunnel in tunnels) {
      final rules = activeRules[tunnel.id] ?? const [];
      files[tunnel.ruleSetFileName] = render(rules, platform: platform);
      files[tunnel.ipRuleSetFileName] = renderIP(rules);
    }
    return files;
  }

  /// IP rules live in route-only twins. Reserved ranges are carved from wide
  /// patterns so they cannot capture Wayfork or unroutable addresses.
  static String renderIP(List<Rule> rules) {
    final ranges = <String>[];
    for (final rule in rules.where((rule) => rule.isIP)) {
      final prefix = IPv4Prefix.parse(rule.pattern);
      if (prefix == null) continue;
      ranges.addAll(
        prefix
            .subtractingAll(RulePattern.reservedRanges)
            .map((value) => value.toString()),
      );
    }
    return '${JsonText.render(<String, Object?>{
      'version': version,
      'rules': ranges.isEmpty ? <Object?>[] : <Object?>[
              <String, Object?>{'ip_cidr': ranges},
            ],
    })}\n';
  }

  /// Domain and application rules are separate objects because sing-box ANDs
  /// different matcher kinds within one object and ORs objects in a set.
  static String render(
    List<Rule> rules, {
    WayforkPlatform platform = WayforkPlatform.windows,
  }) =>
      _render(rules, domains: const [], suffixes: const [], platform: platform);

  /// Built-in local names precede the user's Direct exceptions.
  static String renderDirect({
    List<Rule> exceptions = const [],
    WayforkPlatform platform = WayforkPlatform.windows,
  }) => _render(
    exceptions,
    domains: builtInDirectDomains,
    suffixes: builtInDirectSuffixes,
    platform: platform,
  );

  static String _render(
    List<Rule> rules, {
    required List<String> domains,
    required List<String> suffixes,
    required WayforkPlatform platform,
  }) {
    final domain = [...domains];
    final domainSuffix = [...suffixes];
    final domainRegex = <String>[];
    final processPathRegex = <String>[];
    for (final rule in rules) {
      switch (rule.match) {
        case RuleMatch.exact:
          domain.add(rule.pattern);
        case RuleMatch.suffix:
          domain.add(rule.pattern);
          domainSuffix.add('.${rule.pattern}');
        case RuleMatch.wildcard:
          domainRegex.add(RulePattern.wildcardRegex(rule.pattern));
        case RuleMatch.app:
          processPathRegex.add(platform.appPathRegex(rule.pattern));
        case RuleMatch.ip:
          break;
      }
    }

    final domainRule = <String, Object?>{
      if (domain.isNotEmpty) 'domain': domain,
      if (domainSuffix.isNotEmpty) 'domain_suffix': domainSuffix,
      if (domainRegex.isNotEmpty) 'domain_regex': domainRegex,
    };
    final ruleObjects = <Object?>[
      if (domainRule.isNotEmpty) domainRule,
      if (processPathRegex.isNotEmpty)
        <String, Object?>{'process_path_regex': processPathRegex},
    ];
    return '${JsonText.render(<String, Object?>{'version': version, 'rules': ruleObjects})}\n';
  }
}
